### ISSUE TITLE
Possible issues with reproduction

### ISSUE THREAD

**Issue by Parskatt** [TWJianNuo/DKMReProduction#1]
Issue 1 - Lower LR: 

https://github.com/TWJianNuo/DKMReProduction/blob/d8c1d97cb1bf4f14398892b1b9ec6a69bf601877/train/train_mega_dkm_nnparallel_wvalidation.py#L65-L67

You are using a very low learning rate combined with using a large batch size (36). In our training, we used a batch size of 32 and a learning rate 4 times higher. This is probably causing issues. I suggest you use around 5x the learning rate.

Issue 2 - Possible discrepencies between public repo and private training repo:

I will send you some logs from the run, for reproduction. One thing I noted was that it seems that the run ended before 250k iterations. In fact it seems it only ran for about 215k iterations. In the code base at the time the run was set for 300k iterations with the same scheduler, this means have trained 100k iterations at the intitial learning rate, 100k more with factor 0.2, and only around 20k iterations with the lowest learning rate. It is possible that training for longer with the lowest learning rate is detrimental for generalization.

**Comment by Parskatt** [comment]
[depth_loss_16.csv](https://github.com/TWJianNuo/DKMReProduction/files/9704377/depth_loss_16.csv)
I attached a simple csv with some training logs from the run.

**Comment by TWJianNuo** [comment]
Hi!

In the reproduced version, I am doing the following:
1. I set the learning rate to be 4e-4 during reproduction, running to 200k steps (the provided checkpoint.) During another fresh run (currently running), I aligned with the paper's implementation details and change it to the 1e-4. 
I will change back and re-run it for 300k steps.
2. I can reproduce the testing results using the updated codes and default setting (eval_megadepth1500.py).

I would like to double-check the following facts:

1. Do we need to freeze the batch norm layer in the encoder? I saw it should be done in one answer in the github. But I do not see this is implemented.
2. Will install cupy affect the performance? I installed cupy to eliminate the warning in the "local_corr.py" ("Cupy not found, local correlation will not work"). However, you did not require cupy while creating the environment. I am wondering maybe installing the cupy only accelerate inference speed but will not eliminate some backward gradint? (I will update the answer after a detailed look into this issue.)
3. Will the pytorch environment affect the results? I can align with the environment if you execute:
`conda list -e > requirements_conda.txt`
`python -m pip freeze > requirements_pip.txt`

Thanks!

Shengjie

**Comment by TWJianNuo** [comment]
One minor question. In one sample data of the provided csv file:
76.0,1659778761766,0.287984583659381

It looks like the first "76.0" is step and the last '0.287984583659381' is loss. May I know is middle one "1659778761766" refers to the time?

**Comment by Parskatt** [comment]
Yes, it is step, timestamp, loss

**Comment by Parskatt** [comment]
[requirements_pip.txt](https://github.com/TWJianNuo/DKMReProduction/files/9708105/requirements_pip.txt)

**Comment by TWJianNuo** [comment]
Thanks! Copy this.

**Comment by Parskatt** [comment]
Hi, I recorded the state (only batch size 4), where do you want me to send it?

**Comment by Parskatt** [comment]
https://github.com/Parskatt/storage/releases/download/test/full_state.pt

I put it in github storage.

**Comment by TWJianNuo** [comment]
Thanks for your kind support!

I also got the loss results from the released checkpoint:

Released Checkpoint + Training Script on MegaDepth with Bz 12:
Mean: 0.077425, Std: 0.018784
Released Checkpoint + Training Script on MegaDepth with Bz 3:
Mean: 0.075297, Std: 0.033965

From CSV last 4k steps (last 100 datapoints):
Mean: 0.023265, Std: 0.000689

You can view the image of one screenshot of the tensorboard using this [link](https://drive.google.com/file/d/1zinKGUOWgSeiiZlNiAfOZmiPGIPP0uD_/view?usp=sharing).

**Comment by Parskatt** [comment]
This seems really strange. Might there be some issue with how you have preprocessed the data? I'll check the loss I get as well, but my internal experiments have way lower loss than that (at stride 16).

**Comment by TWJianNuo** [comment]
Sure! I will later report the loss using the input batch you passed me as a cross-validation to eliminate potential issue from the data augmentation part. (The reported results are using training data augmentation. But I did not modify the megadepth dataset related codes.)

**Comment by Parskatt** [comment]
I'm getting the same issue as you.
Also the experiment using the public repo training code is yielding significantly worse results on my internal code, I think there may be a bug somewhere.

**Comment by TWJianNuo** [comment]
I see! Thanks for informing me that!

**Comment by Parskatt** [comment]
Lol I found the bug

**Comment by Parskatt** [comment]
https://github.com/Parskatt/DKM/blob/5f4d03860dfd33a2d9e766fadad0f7720096b1a2/experiments/dkmv2/train_outdoor.py#L43-L45

I accidentally swapped the order of decoder and encoder here.

**Comment by Parskatt** [comment]
I hope this should be enough to fix any reproducing issues. Let me know if any problems persist.

**Comment by TWJianNuo** [comment]
I see! Let me re-run the experiments on my side. 

In addition to that, I have one confusion, thought it might not be a bug. It still worth taking a look. In function "warp_kpts" in "utils.py".

The deph used to compute consistency, i.e.:
w_kpts0_depth & w_kpts0_depth_computed.

The w_kpts0_depth_computed is not projected through the intrinsic matrix:
`    w_kpts0_cam = T_0to1[:, :3, :3] @ kpts0_cam + T_0to1[:, :3, [3]]  # (N, 3, L)`
`    w_kpts0_depth_computed = w_kpts0_cam[:, 2, :]`

Where T_0to1 is only a relative pose (not include intrinsic). Noticable, The source code it get adapted utilize same implementation, making me feel maybe it is my mis-understanding of MegaDepth's Data Format.

**Comment by TWJianNuo** [comment]
Ah, nevermind, this is my misunderstanding. I have insufficient understanding about the projection process. The depth value does not change during the intrinsic projection. Sorry for the incovienience.

**Comment by TWJianNuo** [comment]
Hi, sorry to trouble you again. I think there might still persist an issue:

I upload a new script to align with the pickle file you passed to me. From the experiment, I found the loss can be computed with the exact same number using the provided output "out", i.e.:
`asser (objective(state_dict['out'], state_dict['train_batch']) - state_dict['loss']).abs() < 1e-3`

However, the model outputs are different. I uploaded the code to cross-validated this by running the uploaded script "align_outputs.py". I found the model have an average pixel-wise absolute difference of 0.103621 on dense correspondence modality.

Shengjie

**Comment by Parskatt** [comment]
Hmm, I used my internal code to generate the state, but it really should be the exact same model so Im not sure why it would give different outputs. I suggest we revisit this if it turns out that your re-run experiment also cannot reproduce results.

//Johan

**Comment by TWJianNuo** [comment]
I see. Thanks!
